---
title: Introducing BaseSet
author:
  - Lluís Revilla Sancho
date: '2021-01-14'
slug: introducing-baseset
categories: []
tags:
  - BaseSet
  - Software Peer Review
  - data-munging
  - R
  - packages
  - community
package_version: 0.0.14
description: ""
twitterImg: blog/2021/01/14/introducing-baseset/name-of-image.png
output:
  html_document:
    keep_md: yes
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# Options to have images saved in the post folder
# And to disable symbols before output
knitr::opts_chunk$set(fig.path = "", comment = "", collapse = TRUE)

# knitr hook to make images output use Hugo options
knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "{{<figure src=",
      '"', x, '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}\n"
    )
  }
)

# knitr hook to use Hugo highlighting options
knitr::knit_hooks$set(
  source = function(x, options) {
  hlopts <- options$hlopts
    paste0(
      "```r ",
      if (!is.null(hlopts)) {
      paste0("{",
        glue::glue_collapse(
          glue::glue('{names(hlopts)}={hlopts}'),
          sep = ","
        ), "}"
        )
      },
      "\n", glue::glue_collapse(x, sep = "\n"), "\n```\n"
    )
  }
)
```

On this post I will explain the history behind [BaseSet](https://cran.r-project.org/package=BaseSet "BaseSet on CRAN") then a brief introduction to sets, followed by showing what you can do with [BaseSet](https://docs.ropensci.org/BaseSet/).

## Brief BaseSet history

I work studying diseases to try to find what causes them at a research institute associated with an hospital.
Thanks to recent technological advances we can analyze many things of a single patient's sample.
Having so much information available makes difficult to find the cause of diseases (but is much better than not having enough information!).

In order to make it easier to focus and aggregate knowledge, scientists have been storing what do we know of each gene in several ways.
As more information have been gathered efforts to group these genes started to emerge and being used to focus or discard some genes as cause of different process of interest[^1].
Scientists group the genes according to where they are, what do they do, how they are, with which other genes do they interact, and among these categories there are many different groups.

[^1]: See [this article](https://doi.org/10.1371/journal.pcbi.1002375 "Ten Years of Pathway Analysis: Current Approaches and Outstanding Challenges") reviewing the different methods used.

This has lead to several ways of storing this information, and depending on which database and which method you use you end up with different groups.
This in turn means that we aren't so sure of how to group genes and as new experiments are carried out they might change.

I had previously created a package ([BioCor](https://bioconductor.org/packages/BioCor/ "BioCor: Comparing sets of genes")) to compare these sets and returns a numerical value of how similar two groups of genes are.
But I found out that depending on which database I used I got very different results.
So I wanted to be able to know in which groups a gene was present in several data sources in an easy way and also to unify these database and account for these uncertainty.

So I started experimenting extending a class present on [GSEABase](https://bioconductor.org/packages/GSEABase/) that I used previously to measure and see how could I do that.
I found out that there was also some effort from other people through the [Bioconductor](https://bioconductor.org "Project for high-throughput genomic data") slack channel.
So I got in touch with them and we shared some ideas how to do this.

To explore different ways to do it we end up developing three different packages that aim to do similar things but take different approaches.
At the moment BaseSet is published on CRAN and [BiocSet](https://bioconductor.org/packages/BiocSet/ "BiocSet displays different biological sets in a triple tibble format") published on Bioconductor.
Both packages provides methods to work with sets, the mathematical name of groups of elements.

## Sets

Sets is not only used for tennis games.
In mathematics if you have a group of multiple elements such as `c("apple", "bannana", "pineaple")` it is called a set.
A set is a group of elements, usually just unique elements are used (no `c("apple", "bannana", "pineaple", "apple")` .

To learn and understand the world we group elements together.
Usually it is clear what should be inside a group, like the elements above, all of them are fruits.
Sometimes it is not so clear if an element belongs to a certain group: Are tomatoes fruits?[¹](https://en.wikipedia.org/wiki/Tomato#Fruit_versus_vegetable "Wikipedia page about tomatoes and its classification")

An element can also belong to several groups.
Basketball is a sport, but is also played by a team.
So basketball can be on both groups \[Add image of wikipedia classification for Basketball\].
Some say that tomatoes should be both considered as fruits and as vegetables, so they are also on both sets.
We can do this for all elements we can think, pencils, mathematical operations, numbers, words, genes...

When we have more than one set we might be interested in which elements are in two groups, which fruits are also vegetables, or which fruits are not vegetables.
These operations are called set operations.
Every time you have used `%in%`, `merge` in your code you have been using a set operation.
There are many, set operations, the most used are intersection, for which R has the `intersect` and the union in R is done with `c` on vectors.
The equivalents on data.frames are done with `merge` and their arguments (if you are familiar with the `*_join` verbs from dplyr that is one of the reasons why there are so many `*_join` functions).

As you see R already provides some functions to work with sets.
So why writing a new package for sets?
Let's see it with some code:

## Using BaseSet

Now let's see how do we do with BaseSet:

```{r loading}
library("BaseSet")
packageVersion("BaseSet")
```

In order to work with the sets the package defines its own class named TidySet, we can transform our data with `tidySet:`

```{r fruits}
sets <- list(fruits = c("apple", "bannana", "pineaple"))
TS <- tidySet(sets)
TS
```

We can explore some properties of this set:

```{r properties}
nElements(TS)
nSets(TS)
```

then add more fruits

```{r orange}
TS1 <- add_elements(TS, "orange")
nElements(TS1)
TS1
```

We can see it has one more element but it isn't displayed, that's is because we didn't record to what set this element belongs to:

```{r new-relation}
relation <- data.frame(sets = "fruits", elements = "orange")
TS1 <- add_relation(TS, relation)
TS1
```

As you have noticed there is a third column named fuzzy that so far is always 1.
This column is a value between 0 and 1 with the membership of an element to that set.
1 is that it is fully and 0 it is that there isn't any relationship.
Values between encode the uncertainty.

```{r tomatoes}
relation <- data.frame(sets = c("vegetables", "fruits"), 
                       elements = c("tomatoes", "tomatoes"))
TS2 <- add_relation(TS1, relation)
TS2
```

Now that we have two sets we can perform more operations:

```{r intersection}
intersection(TS2, c("vegetables", "fruits"))
union(TS2, c("vegetables", "fruits"))
```

Note that the names of the resulting sets have been automatically set up using the [appropriate symbols](https://www.mathsisfun.com/sets/symbols.html "Link to a website with some set operations symbols").

So far we have set up how to count and merge different sets but one of the goals of BaseSet is to also encode the uncertainty when we assign an element to a set.

To account for uncertainty mathematicians have come with the idea of fuzzy sets.
Fuzzy sets are sets but instead of knowing for sure that an element belongs to the set, there is a value, the fuzzy value or membership, that measures the membership of the element to a set.

\[Find a layman example of fuzzy set, popcorn or something\]

That fuzzy value can be a probability, so if researchers perform experiments and 8/10 times the elements does a process we could say that the fuzzy value is 0.8.
This helps to compare how sure we are of our own knowledge.

```{r fuzzy}

```

## Conclusions

I [submitted](https://github.com/ropensci/software-review/issues/359) this package to [rOpenSci](https://ropensci.org "rOpenSci") because I wanted more feedback on these fuzzy sets methods and wasn't sure of the mix I made of S4 methods and S3 methods.

The topic is not a common one so although I submitted the package on January for review it started on June (after 2 months of hiatus due to COVID).
This long waiting time was because it was hard for the editor (thanks Anna Krystalli) to find two reviewers with all that was going on.

Once two reviewers agreed the feedback started coming (Many thanks Zebulun Arendsee and Jennifer Chang!).
The reviewers pointed out inconsistencies on the order of the output, suggested more methods, pointed out to mismatches between the documentation and the code.

Most importantly much of the comments and exchange was around some methods and explanations for fuzzy sets were (are?) not clear.
Mainly if the fuzzy values, also called memberships are probabilities or not.

I made my best efforts to improve the package and hope you'll find them useful in your projects.

Citation of the primary literature[^2].

[^2]: Sciaini, M., Fritsch, M., Scherer, C., & Simpkins, C. E.
    (2018).
    NLMR and landscapetools: An integrated environment for simulating and modifying neutral landscape models in R.
    Methods in Ecology and Evolution, 9(11), 2240-2248.
    <https://doi.org/10.1111/2041-210X.13076>

Citation of a website[^3].

[^3]: Elin Waring, Michael Quinn, Amelia McNamara, Eduardo Arino de la Rubia, Hao Zhu and Shannon Ellis (2019).
    skimr: Compact and Flexible Summaries of Data.
    R package version 1.0.7.
    <https://CRAN.R-project.org/package=skimr>

Citation of an R package[^4].

[^4]: Hugo static site generator.
    <https://gohugo.io/>

### Subsection heading

We recommend the use of [Hugo shortcodes](https://gohugo.io/content-management/shortcodes/) to include images, tweets, videos, gists, etc.
In the Rmd file they need to be between `<!--html_preserve-->` and `<!--/html_preserve-->`.

**Add an image** by using a Hugo shortcode.
The image is saved under `/content/blog/YYYY-MM-DD-slug/name-of-image.png`.

<!--html_preserve-->
{{< figure src = "name-of-image.png" width = "400" alt = "this is the alternative text" >}}
<!--/html_preserve-->

Consult the Technical Guidelines for tips on changing image size, alignment, and for advice on alternative text.

Below is another image, a plot generated by a code chunk, so you might see how to add alternative text and other options in this case.
You can also use [all options documented for the figure shortcode](https://gohugo.io/content-management/shortcodes/#figure) except for `src` that will be created automatically when knitting.

```{r chunkname, hugoopts=list(alt="alternative text please make it informative", caption="this is what this image shows, write it here or in the paragraph after the image as you prefer", width=300)}
plot(1:10)
```

Once this file is knitted the plot above will be inserted with the correct syntax.

**Embed a tweet** by using a Hugo shortcode.

<!--html_preserve-->
{{< tweet 1138216112808529920 >}}
<!--/html_preserve-->

**Tweak code highlighting** Below you can see how to use Hugo [Chroma code highlighting options](https://gohugo.io/content-management/syntax-highlighting/#highlight-shortcode): highlight the 1st and 4th to 6th code lines, number lines as a table starting from 199.

```{r hl, hlopts = list(linenos='table',hl_lines='[1,"4-6"]',linenostart=199)}
a <- 1
a <- 1
a <- 1
a <- 1
a <- 1
a <- 1
a <- 1
```

**Add citation or footnote** text by using the format below
